### PR TITLE
Implement API list and delete

### LINKED ISSUE #1630: Support to delete and list existing APIs via the adapter REST API
### Describe your problem(s)
To delete API and  list APIs using apictl Command Line tool

### Describe your solution
Implement delete API and  list APIs via the REST API

### PR DESCRIPTION
**Pull request by SuKSW**
### Purpose
Implementation to delete an API and to list existing APIs via the REST API, which will be done using apictl.
This PR was created after rebasing the original PR [here](https://github.com/wso2/product-microgateway/pull/1631)

### Issues
Fixes #1630 

### Automation tests
 - Unit tests added: Yes/No
 - Integration tests added: Yes/No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: adapter/pkg/xds/xds_server.go [hunk 1]
```diff
@@ -763,3 +789,41 @@ func stopConsulDiscoveryFor(clusterName string) {
 		close(doneChan)
 	}
 }
+
+// ListApis returns a list of objects that holds info about each API
+func ListApis(apiType string, limitP *int64) *apiModel.APIMeta {
```
Comment (praminda) [inline adapter/pkg/xds/xds_server.go:794]:
```suggestion
func ListApis(apiType string, limit *int64) *apiModel.APIMeta {
```

File: adapter/resources/adapterAPI.yaml [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+        type: array
+        items:
+          type: string
+  DeployResponse:
+    type: object
+    properties:
+      action:
+        type: string
+      info:
+        type: string
```
Comment (praminda) [inline adapter/resources/adapterAPI.yaml:297]:
```suggestion
        type: string

```

**Review by praminda — CHANGES_REQUESTED** [review]

File: adapter/pkg/api/apis_impl.go [hunk 3]
```diff
@@ -44,26 +45,19 @@ const (
...
-// content.
+// extractAPIProject accepts the API project as a zip file and returns the extracted content
 // The apictl project must be in zipped format. And all the extensions should be defined with in the openAPI
 // definition as only swagger.yaml is taken into consideration here. For websocket APIs api.yaml is taken into
 // consideration. API type is decided by the type field in the api.yaml file.
-func ApplyAPIProject(payload []byte, envrionments []string) error {
-	if len(envrionments) == 0 {
-		envrionments = append(envrionments, "Production and Sandbox")
-	}
+func extractAPIProject(payload []byte) (apiJsn []byte, swaggerJsn []byte, upstreamCerts []byte,
```
[the remark this answers is not in the document]
Comment (SuKSW) [reply adapter/pkg/api/apis_impl.go:52]: That'll reduce space for errors as well, adding a struct.

File: adapter/pkg/api/apis_impl.go [hunk 4]
```diff
@@ -100,36 +94,112 @@ func ApplyAPIProject(payload []byte, envrionments []string) error {
...
+	if err != nil {
+		return err
+	}
+	xdsUpdateAPI(name, version, apiType, apiJsn, swaggerJsn, upstreamCerts, environments)
+	return nil
+}
+
+// ApplyAPIProjectWithOverwrite is called by the rest implementation to differentiate
+// between create and update using the overwrite param
+func ApplyAPIProjectWithOverwrite(payload []byte, environments []string, overwriteP *bool) error {
```
[the remark this answers is not in the document]
Comment (SuKSW) [reply adapter/pkg/api/apis_impl.go:143]: Updating this to a meaningful name

File: adapter/pkg/api/apis_impl.go [hunk 5]
```diff
@@ -100,36 +94,112 @@ func ApplyAPIProject(payload []byte, envrionments []string) error {
...
+	}
+	if apiType == mgw.HTTP {
+		xds.UpdateAPI("default", name, version, apiType, swaggerJsn, upstreamCerts, environments)
+	} else if apiType == mgw.WS {
+		xds.UpdateAPI("default", name, version, apiType, apiJsn, upstreamCerts, environments)
+	}
+}
+
+// DeleteAPI calls the DeleteAPI method in xds_server.go
+func DeleteAPI(apiName string, version string, vhostP *string) error {
```
[the remark this answers is not in the document]
Comment (SuKSW) [reply adapter/pkg/api/apis_impl.go:184]: Updating this to vhost

File: adapter/pkg/api/apis_impl.go [hunk 6]
```diff
@@ -100,36 +94,112 @@ func ApplyAPIProject(payload []byte, envrionments []string) error {
...
+	}
+	if !overwrite && exists {
+		loggers.LoggerAPI.Infof("Error creating new API. API %v:%v already exists.", name, version)
+		return errors.New(mgw.AlreadyExists)
+	}
+	xdsUpdateAPI(name, version, apiType, apiJsn, swaggerJsn, upstreamCerts, environments)
 	return nil
 }
 
+func xdsUpdateAPI(name, version, apiType string, apiJsn, swaggerJsn, upstreamCerts []byte,
```
[the remark this answers is not in the document]
Comment (SuKSW) [reply adapter/pkg/api/apis_impl.go:171]: Renaming this

File: adapter/pkg/api/apis_impl.go [hunk 7]
```diff
@@ -100,36 +94,112 @@ func ApplyAPIProject(payload []byte, envrionments []string) error {
...
+func DeleteAPI(apiName string, version string, vhostP *string) error {
+	if vhostP == nil || *vhostP == "" {
+		vhost := "default"
+		vhostP = &vhost
+	}
+	return xds.DeleteAPI(apiName, version, *vhostP)
+}
+
+// ListApis calls the ListApis method in xds_server.go
+func ListApis(apiTypeP *string, limitP *int64) *apiModel.APIMeta {
```
[the remark this answers is not in the document]
Comment (SuKSW) [reply adapter/pkg/api/apis_impl.go:193]: Adding meaningful names

File: adapter/pkg/api/restserver/embedded_spec.go [hunk 8]
```diff
@@ -60,21 +61,21 @@ func init() {
   "host": "apis.wso2.com",
   "basePath": "/api/mgw/adapter/0.1",
   "paths": {
-    "/apis": {
+    "/api": {
```
[the remark this answers is not in the document]
Comment (SuKSW) [reply adapter/pkg/api/restserver/embedded_spec.go:64]: Updating this to /apis as discussed

File: adapter/pkg/xds/xds_server.go (hunk 1 above)
Comment (SuKSW) [reply adapter/pkg/xds/xds_server.go:794]: Updating this

File: adapter/resources/adapterAPI.yaml (hunk 2 above)
Comment (SuKSW) [reply adapter/resources/adapterAPI.yaml:297]: adding space at the end of file

File: adapter/pkg/api/apis_impl.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (SuKSW) [reply adapter/pkg/api/apis_impl.go:52]: Using the name ProjectAPI to pass lint errors

**Comment by SuKSW** [comment]
Closing PR in order to resolve conflicts.

**Pull request closed without merging**